src/Features/CSharp/Portable/Formatting/CSharpFormattingInteractionService.cs

```diff
@@ -0,0 +1,463 @@
+                return false;
+            }
+
+            if (ch == ';' && !options.GetOption(FormattingOptions2.AutoFormattingOnSemicolon, LanguageNames.CSharp))
+            {
+                return false;
+            }
+
+            // don't auto format after these keys if smart indenting is not on.
+            if ((ch == '#' || ch == 'n') && !smartIndentOn)
+            {
+                return false;
+            }
+
+            return _supportedChars.Contains(ch);
+        }
+
+        public async Task<IList<TextChange>> GetFormattingChangesAsync(
+            Document document,
+            TextSpan? textSpan,
+            DocumentOptionSet? documentOptions,
+            CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var span = textSpan ?? new TextSpan(0, root.FullSpan.Length);
+            var formattingSpan = CommonFormattingHelpers.GetFormattingSpan(root, span);
+            var options = documentOptions;
+            if (options == null)
+            {
+                var inferredIndentationService = document.Project.Solution.Workspace.Services.GetRequiredService<IInferredIndentationService>();
+                options = await inferredIndentationService.GetDocumentOptionsWithInferredIndentationAsync(document, explicitFormat: true, cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
+            return Formatter.GetFormattedTextChanges(root,
+                SpecializedCollections.SingletonEnumerable(formattingSpan),
+                document.Project.Solution.Workspace, options, cancellationToken);
+        }
+
+        public async Task<IList<TextChange>> GetFormattingChangesOnPasteAsync(
+            Document document,
+            TextSpan textSpan,
+            DocumentOptionSet? documentOptions,
+            CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var formattingSpan = CommonFormattingHelpers.GetFormattingSpan(root, textSpan);
+            var service = document.GetLanguageService<ISyntaxFormattingService>();
+            if (service == null)
+            {
+                return SpecializedCollections.EmptyList<TextChange>();
+            }
+
+            var rules = new List<AbstractFormattingRule>() { new PasteFormattingRule() };
+            rules.AddRange(service.GetDefaultFormattingRules());
+
+            var options = documentOptions;
+            if (options == null)
+            {
+                var inferredIndentationService = document.Project.Solution.Workspace.Services.GetRequiredService<IInferredIndentationService>();
+                options = await inferredIndentationService.GetDocumentOptionsWithInferredIndentationAsync(document, explicitFormat: false, cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
```

Comment:
This was changed from a simple null-coalescing to this `if`, to extract out getting the `inferredIndentationService`.